ant-cli/src/actions/download.rs

```diff
@@ -7,123 +7,192 @@
+            println!("Try the `file list` command to get addresses you have access to");
+            process::exit(exit_code::INVALID_INPUT_EXIT_CODE);
+        }
+    };
+
+    // wait for upload to complete
+    if let Err(e) = stop_summary_collection.send(()) {
+        error!("Failed to send upload completed event: {e:?}");
     }
+
+    let task_summary = task_summary_thread.await?;
+    let download_output = FileDownloadOutput {
+        exit_code: result
+            .as_ref()
+            .map_or_else(|(err, _)| exit_code::download_exit_code(err), |_| 0),
+        addr: addr.to_string(),
+        public,
+        task_summary,
+        start_time,
+        end_time: SystemTime::now(),
+        failed_files: result
+            .as_ref()
+            .map_or_else(|(_, failed_files)| failed_files.clone(), |_| vec![]),
+    };
+
+    if json {
+        download_output.print_json();
+    } else {
+        download_output.print();
+    }
+
+    if let Err((err, _)) = result {
+        process::exit(exit_code::download_exit_code(&err));
+    }
+    Ok(())
 }
 
 async fn download_private(
-    addr: &str,
     private_address: PrivateArchiveAccess,
     dest_path: &str,
     client: &mut Client,
-) -> Result<()> {
+) -> Result<(), (DownloadError, Vec<(String, String)>)> {
     let archive = client
-        .archive_get(private_address)
+        .archive_get(private_address.clone())
         .await
-        .wrap_err("Failed to fetch data from address")?;
+        .map_err(|err| {
+            error!("Failed to fetch archive from address: {private_address:?}");
+            (err.into(), vec![])
+        })?;
 
-    let progress_bar = get_progress_bar(archive.iter().count() as u64)?;
+    let progress_bar = get_progress_bar(archive.iter().count() as u64)
+        .inspect_err(|err| {
+            error!("Failed to create progress bar: {err}");
+        })
+        .ok();
     let mut all_errs = vec![];
+    let mut last_error = None;
+
     for (path, access, _meta) in archive.iter() {
-        progress_bar.println(format!("Fetching file: {path:?}..."));
         let bytes = match client.data_get(access.clone()).await {
             Ok(bytes) => bytes,
             Err(e) => {
-                let err = format!("Failed to fetch file {path:?}: {e}");
-                all_errs.push(err);
+                all_errs.push((path.to_string_lossy().into_owned(), format!("{e}")));
+                last_error = Some(e);
                 continue;
             }
         };
 
         let path = PathBuf::from(dest_path).join(path);
         let here = PathBuf::from(".");
         let parent = path.parent().unwrap_or_else(|| &here);
-        std::fs::create_dir_all(parent)?;
-        std::fs::write(path, bytes)?;
-        progress_bar.clone().inc(1);
+        std::fs::create_dir_all(parent).map_err(|err| {
+            error!("Failed to create parent directories for {path:?}: {err}");
```

Comment:
Use `inspect_err`

Reply:
Same as above, we actually map the error to a different type here, so it should be fine for the logging to go in here.